src/Algebra/Graph/Bipartite/AdjacencyMap.hs

```diff
@@ -650,6 +661,116 @@ vertexSet = Set.fromAscList . vertexList
 edgeSet :: (Ord a, Ord b) => AdjacencyMap a b -> Set.Set (a, b)
 edgeSet = Set.fromAscList . edgeList
 
+data Part = LeftPart | RightPart
+    deriving (Show, Eq)
+
+otherPart :: Part -> Part
+otherPart LeftPart  = RightPart
+otherPart RightPart = LeftPart
+
+type PartMap a = Map.Map a Part
+type PartMonad a r = MaybeT (State (PartMap a)) r
+
+-- | An odd cycle.
```

Comment:
I think it's worth clarifying how a cycle is represented. For example, a cycle `1 -> 2 -> 3 -> 1` is represented by `[1, 2, 3]`.

Reply:
...or it could be `[2,3,1]` as well.

Reply:
It should be vice verse — [1, 2, 3] represents 1 -> 2 -> 3 -> 1.

Reply:
> It should be vice versa — [1, 2, 3] represents 1 -> 2 -> 3 -> 1.

Yes, this formulation seems better.